### DISCUSSION TITLE
2024-02-09 - How to read source code: Interfaces and Abstract Classes - Chewing the Fat 📈

### DISCUSSION THREAD

**Discussion by piers-sinclair** [SSWConsulting/SSW.Rules.Content#7930] [Chewing the Fat]
Hey SSWers,

Let's talk about How to read source code: Interfaces and Abstract Classes!

https://www.ssw.com.au/rules/read-source-code/

**Comment by GordonBeeming** [comment]
I loved the video, nice 1 @AttackOnMorty. Funny enough I was planning on doing the same 1 and then I saw yours go live 😅

**Comment by BrookJeynes** [comment]
Great video, very easy to follow and understand.

**Comment by bahjatalaadelssw** [comment]
Nice work @AttackOnMorty , very clear and easy to follow. Good size for quick learning!

**Comment by jakebayliss** [comment]
I like the series style - excited for more!

**Comment by drwharris** [comment]
Love the idea of a beginner series. Well done 👍

PS: Love the Monkey :)
![image](https://github.com/SSWConsulting/SSW.Rules.Content/assets/19632724/bfa6c011-373a-4216-b134-8ea42cdfa293)

**Comment by tkapa** [comment]
Pretty cool video, I think it's odd to start with Interfaces and Abstract classes. They aren't typically the barrier I think of when it comes to teaching new people how to read code.

**Comment by ricksu978** [comment]
Looking forward to diving into more topics on your next episode. Keep up the great work! 👍

**Comment by ozdemir-mehmet-cer** [comment]
AAA+++ will watch again

**Comment by UlyssesMaclaren** [comment]
Am I lost? How did I find myself here! I'm scared and alone in a sea of NERDS!!! 
lol good video, Luke. with a little extra ChatGPT-ing, i think I understand what you were talking about now.

**Comment by YalalCakirSSW** [comment]
Great video. Easy to follow with good examples. I would like to see more educational coding videos.

**Comment by joshbermanssw** [comment]
The examples delivered the lesson well! I'm a fan of the educational series!

**Comment by zacharykeeping** [comment]
Good video - looking forward to more in this style.

**Comment by MarcusIrmscherSSW** [comment]
I'm with Uly, this was mostly over my head but great video. Very well presented Like, awesome job 🏆

**Comment by jeoffreyfischer** [comment]
I find educational videos incredibly engaging and effective. They provide clear explanations and use real-life examples to make complex concepts easy to understand.

**Comment by chrisschultzssw** [comment]
As a SysAdmin... I think I understood the content 🤔.
Great video though!

**Comment by TobyChurches** [comment]
Great video! Very well presented! Its these sorts of videos which I loved in the early years of Uni! I think educational series would be great for SSW.TV 🎥

**Comment by Calinator444** [comment]
I like the video. If you wanted a more language independent way of demonstrating how interfaces work using it might be a good idea to use psuedocode or class diagrams. Although I think the examples you provided were simple enough that you don't really need to know typescript to understand the examples.

**Comment by christoment** [comment]
Great video and rule - especially for fresh developers 😄 

My take on the video: since the rule already covers the explanation in the video, we should've tune the video content so non-devs can also understand the concept 😄

**Comment by Marxoz** [comment]
I love these kind of videos. Great information, and straight to the point.

**Comment by AntPolkanov** [comment]
Love educational videos!
The topic is not an easy one. Over the last couple of years the difference between interfaces and abstract classes has become hazy after some languages started adding default method implementations for interfaces (e.g. C# 8 and Java 8).

**Comment by isaaclombardssw** [comment]
Education series is a great concept, would help more junior devs (like myself) to get to grips with some harder topics. 

As for feedback for the specific video, I think that discussing interfaces/abstract classes is somewhat of a weak link to reading source code (it's definitely a part of it, but the video title is a bit misleading when it mostly talks about those two concepts). The video itself is quite useful as a interface/abstract class revision/lesson.

**Comment by tombui99** [comment]
This is an awesome video especially for fresh developers when starting a new project.

I'm very looking forward to more educational contents like this one.

**Comment by SamBWagner** [comment]
Educational videos are THE WAY! This was super cool! 

tip: I'd always show the proof of the pudding. In the video, we showed TimePro as the example but then never delved into the source code to show HOW one would understand.

Also, we should 100% recommend a project for each one of these videos so people can go and learn in a hands on way.

**Comment by Hona** [comment]
> Interfaces and Abstract classes - Any war stories?

Interfaces that do too much. But also when they do SRP too much and theres 999k interfaces on a class. Just be pragmatic. Too much common logic in some Abstract base class that everything uses but some of that 'base' logic is only needed for 1 inheritor (therefore making the inheritor more confusing & the abstract class more tangled)

> Interfaces and Abstract classes - Any tips for understanding the source code?

Interfaces & Abstract classes won't show you the code. You have to accept not knowing anything as well as not needing to understand code/all paths fully to move on. You need to navigate to usages/definitions/implementations quickly & jump through an end to end path you're trying to understand. Interfaces & Abstract classes are just a small thought in that larger skill set. 

> Interfaces and Abstract classes - Any tips for identifying or fixing problems?

Same as what I said above. Move quickly, don't get caught up on understanding. Code is sequential & logical. Use a debugger. Everything can be black box tested, which fundamentally lets you scope smaller and smaller pieces of code to identify the root cause of an issue.

--- 

How do you all deep dive code jungles? 🌲🎋🔪🌴🎄🌳

**Comment by AshAnil** [comment]
Nice video ⭐

**Comment by ncn-ssw** [comment]
Good video for junior coders. I had not really stopped to think about how the first thing I've done in the past to fix a sub-optimal codebase is fix its data structures.

**Comment by leahy268** [comment]
Somewhere in my deep dark history (uni days) I vaguely remember abstract classes.
Good video - Not sure how much I really understood it 😊

**Comment by jimmy1992abc** [comment]
Nice video, well structured, and the rule is excellent.
I believe this video can get more hits because the takeaway is straightforward.

**Comment by synnoveisabel** [comment]
Great video Luke! It was structured very well so it was easy to follow 🚀

**Comment by ozair-ashfaque** [comment]
Awesome video, Luke! Thanks for breaking down interfaces and abstract classes in such a straightforward manner!

**Comment by Dhruv-0987** [comment]
Amazing video Luke, I would prefer to see videos like this every week, simplifying a complicated programming topic to an easy one.

**Comment by amankumarrr** [comment]
Cool video Luke, well structured and clearly explained.

**Comment by vladislav-kir** [comment]
Awesome video from Luke! I would love to see more short educational videos :)

**Comment by JackDevAU** [comment]
Luke is a good presenter but I think we need to either put this into a series of beginner videos and label it correctly. 
Wasn't sure why we had a section in the video showing TimePro's Invoices page... 
Something like this but shorter, on TikTok would be great content. Something like, hey heres an interface. 
Heres how you use it. 
Heres how you can work with it. 
Quick, simple and awesome ⭐ 

but great work Luke M 💪💪

**Comment by JeanThirion** [comment]
Moar educational videos !!

**Comment by bettybondoc** [comment]
I love this video format/style! Turning a complex topic into something even non-devs can understand is a mark of a great presenter. I love how the examples are kept simple and not intimidating—adding just what's needed to explain the concept.

**Comment by KahaMason** [comment]
Awesome video! Having an educational series is a great idea to hone the skills of new software developers. 😊

**Comment by PennyWalker** [comment]
Great first video Luke! Well explained, I enjoyed watching it. 🥳

**Comment by stephanfakossw** [comment]
It's a great video Luke ⭐ 
Do more...

**Comment by danielmackay** [comment]
Great video @AttackOnMorty.  I liked the way you presented the content.

As for the rule itself, although interfaces and abstract classes can play a part in understanding a code base, I don't believe they are the golden ticket.  There may be large portions of the codebase that don't use either of these, and that's fine.

When trying to understand a system and the code within, I find the best place to start is zoomed at a '1000ft view'.  Get an understanding of a system and it's dependencies with external systems.  Then zooming into each system to understand the big areas of functionality, then zooming into each one of those for more detail.

I also would have like to have seen the Dependency Inversion principle mentioned as a benefit of using interfaces.

**Comment by tomek-i** [comment]
I think the video needs a little bit more detail about the differences and maybe 1-2 examples to drive it home.

**Comment by jerryluossw** [comment]
Great rule, very good summary

**Comment by Aibono1225** [comment]
Love the video, very easy to understand!

**Comment by johnxu16** [comment]
Great video! But if there is a real world case, the content would be much more persuasive.

**Comment by sethdaily** [comment]
A bit out of my current knowledge 🤣

**Comment by cvionnet** [comment]
Great video Luke! It's nice to see more educational videos 👍

**Comment by AlexRothwell** [comment]
This was a great intro video on the topic and I could see a series like this being quite useful for something like FireBootCamp

**Comment by MattParkerDev** [comment]
Would think interfaces and classes would be better demonstrated in C#

**Comment by Ben0189** [comment]
Great video Luke ! good clarification in between interface and abstract class. From my understanding interfaces also widely use in Dependency injection where the class used a services (e.g. controller) have no idea on which concrete implementation is injected by using an interface. This make the code easier to maintain (you can easily swap the implementation without breaking other parts of references)

**Comment by matt-goldman** [comment]
Great video, I would love to see more content like this. Would be good for it to form part of a playlist too, possibly even become part of FBC.

**Comment by AttackOnMorty** [comment]
Interfaces and abstract classes are not the starting point of reading source code. It's one of the important parts you need to pay attention to. For how to read the source code in full, check out this rule: https://www.ssw.com.au/rules/read-source-code/

**Comment by 2fernandez** [comment]
What a fantastic video by Luke ❤️.
We need so much more of this style of video that actually walks through code and adds value to anyone learning code.

**Comment by tino-liu** [comment]
It definitely helps in many cases with pinpointing where the business logic is at and how it is implemented when working on huge projects.

**Comment by PothieuG** [comment]
It was a nice video, Luke explained very well, with an intelligible flow !

**Comment by jimmidier** [comment]
Great work Luke, love to see more videos in this style!

**Comment by babakamyljanovssw** [comment]
Amazing video Luke! Would love to see more educational videos in the future.

**Comment by jackreimers** [comment]
Great video, I think it would be very helpful when learning about interfaces and abstract classes for the first time!
